p2p/server.py

```diff
@@ -0,0 +1,156 @@
+import asyncio
+import os
+import logging
+import secrets
+import sys
+
+from typing import Tuple, cast
+from eth_keys import (
+    datatypes,
+    keys,
+)
+from eth_utils import (
+    decode_hex,
+    keccak,
+)
+
+from p2p.auth import (
+    HandshakeResponder,
+    decode_auth_plain,
+    decode_auth_eip8,
+)
+from p2p.cancel_token import CancelToken
+from p2p.constants import (
+    ENCRYPTED_AUTH_MSG_LEN,
+    HASH_LEN,
+)
+from p2p.ecies import ecdh_agree
+from p2p.exceptions import OperationCancelled
+from p2p.kademlia import (
+    Address,
+    Node,
+)
+from p2p.peer import PeerPool, BasePeer, PeerPoolSubscriber, ETHPeer, LESPeer
+from p2p.utils import sxor
+
+
+class Server(PeerPoolSubscriber):
+    """Server listening for incoming connections"""
+    logger = logging.getLogger("p2p.server.Server")
+    
+    def __init__(self, privkey: datatypes.PrivateKey, server_addr: Tuple[str, str], peer_pool: PeerPool) -> None:
+        self.cancel_token = CancelToken('Server')
+        self.incoming_connections = []
+        self.peer_pool = peer_pool
+        self.peer_pool.subscribe(self)
+        self._running_peers = set()
+        self.privkey = privkey
+        self._server_address = server_addr
+
+    def register_peer(self, peer: BasePeer) -> None:
+        asyncio.ensure_future(self.handle_peer(cast(ETHPeer, peer)))
+
+    async def handle_peer(self, peer: ETHPeer) -> None:
+        """Handle the lifecycle of the given peer."""
+        self._running_peers.add(peer)
```

Comment:
@gsalgado Am I missing some steps here that are needed to add a peer to the peer_pool?

Reply:
@njgheorghita I *think* that the `PeerPool` class is going to need a new API for this.  It looks like the peer needs to be added to `PeerPool.connected_nodes`, however, we should not directly push the peer into that data structure.  You can probably do something as simple as:

```python
class PeerPool:
    ...
    def add_peer(self, peer):
        self.connected_nodes[peer.remote] = peer
        for subscriber in self._subscribers:
            subscriber.register_peer(peer)
```

Then, from within your `receive_handshake` method, instead of calling `self.register_peer(...)` you would just instead call `peer_pool.add_peer(peer)` with the knowledge that it will *subscribe* you to the peer.  Similarly, you'll need to implement `Server.remove` to handle the other call from the peerpool when the peer is removed.